### PR TITLE
Shopper → Mini Cart → Can view translated mini cart contents block

### LINKED ISSUE #5814: Cover Mini Cart block with tests
Recently, @nielslange has identified several critical flows for the Cart block that they would like to cover with automated testing. This issue is to evaluate if we would like to have the same tests for the Mini Cart block:

* https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/5751
* https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/5752
* https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/5753
* https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/5754
* https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/5755
* https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/5756
* https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/5762
* https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/5763

**Important:** take into account the issues above are for the Cart block. We should create separate ones for the Mini Cart block.

There are a couple more that come to my mind:

* Shopper → Mini Cart → Can open/close the drawer
* Shopper → Mini Cart → Can go to Cart page

We already have some tests for the Mini Cart block that can be found here:

* https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/assets/js/blocks/cart-checkout/mini-cart/test/block.js
* https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/tests/e2e/specs/backend/mini-cart.test.js

### LINKED ISSUE #5831: Critical flows: Shopper → Mini Cart → Can view translated mini cart contents block
### Note

Create e2e tests to verify that a shopper can see the translated Mini Cart Contents.

### File

`tests/e2e/specs/shopper/mini-cart.test.js`

Part of #5814

### PR DESCRIPTION
**Pull request by dinhtungdu**
Fixes #5831, also closes #5814

### Notes

This PR adds e2e tests to verify that a shopper can see the translated Mini Cart Contents.

I used `nl_NL` instead of `fr_FR` that @alexflorisca used in #6101 because it contains more up-to-date translations at the time of writing.

~@alexflorisca I download the translation at the setup phase to make the test more reliable.~ Update: the flaky tests are caused by different issue, the translation test itself is reliable.

I also updated `shopper.block.emptyCart()` to make it language agnostic.

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [x] E2E tests

### Manual Testing

1. Check out this PR.
2. Start Docker Desktop.
3. Run `npm run wp-env destroy && npm run wp-env start` to start a fresh e2e environment.
4. Run `npm run test:e2e -- -t 'Translated Mini Cart' -i 'tests/e2e/specs/shopper/mini-cart.test.js'` to run only this test case.
5. Run `npm run test:e2e` to run the whole test suite.

### CODE REVIEW

File: bin/wp-env-config.sh [hunk 1]
```diff
@@ -20,4 +20,13 @@ wp plugin list
 wp theme activate storefront
 wp wc customer update 1 --user=1 --billing='{"first_name":"John","last_name":"Doe","company":"Automattic","country":"US","address_1":"addr 1","address_2":"addr 2","city":"San Francisco","state":"CA","postcode":"94107","phone":"123456789"}' --shipping='{"first_name":"John","last_name":"Doe","company":"Automattic","country":"US","address_1":"addr 1","address_2":"addr 2","city":"San Francisco","state":"CA","postcode":"94107","phone":"123456789"}'
 
+
+## Prepare translation for the test suite
+wp language core install nl_NL
```
Comment (alexflorisca) [inline bin/wp-env-config.sh:24]: When discussing how to best approach translations, @nielslange and I talked about doing this and decided against it. Instead I install and activate the language in a before hook in the test that needs it, and uninstall after the test runs. The reason being that we wanted to encapsulate the tests as much as possible. We wanted to avoid adding anything here that would only be used by 1 or 2 tests. Although thinking about it now, if you guys are going to write translation tests for all the components, and use the same language, then it definitely makes sense to put them here. If we decide to go with this approach, I'll edit the `cart-checkout-translations` to use dutch as well.

Comment (dinhtungdu) [reply bin/wp-env-config.sh:24]: > The reason being that we wanted to encapsulate the tests as much as possible. We wanted to avoid adding anything here that would only be used by 1 or 2 tests.

@alexflorisca I totally agree with this. But looks like we're having flaky tests by installing the language from the test (Cart translation E2E tests are failed in all of my PRs, this PR too). I think installing the language in `wp-env-config.sh` is more reliable and predictable. Even if it's not a nice solution, it works most reliably for me. Also, the translation doesn't affect other tests because they all use English.

Comment (nielslange) [reply bin/wp-env-config.sh:24]: @dinhtungdu & @alexflorisca Another way of encapsulating the languages and only activating/deactivating for certain tests, could be to mock a plugin. When facing problems implementing https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/6211, @PanosSynetos suggested using a helper plugin in pca54o-3k7-p2#comment-3880. 

Instead of installing the language in `wp-env-config.sh`, we could create a mock plugin, e.g. `/tests/mocks/woo-switch-language`, and then load this plugin in `.wp-env.json` using the following code:

```json
	"plugins": [
		"./tests/mocks/woo-switch-language",
	],
```

That language switch plugin itself would look for a URL parameter, e.g. `?switch_language=nl_NL`. Once this URL parameter gets called, we can then change the site language and fetch the translations for the plugin using WordPress core functionality and/or WP-CLI.

Not saying that this is the best approach, but I wanted to share this alternative approach with you.

Oh. and on https://github.com/nielslange/woo-test-helper you can find the helper plugin I created for https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/6211/files. For today, I planned to add this plugin as mock plugin to Woo Blocks, similar to the mock themes as seen here:

https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/cdc3138ffbe14cc047c1ba694e9db6b9f76ba827/.wp-env.json#L8-L12

Comment (alexflorisca) [reply bin/wp-env-config.sh:24]: Makes sense, I think CLI is a bit cleaner if we can use that, although I know in your case @nielslange it didn't work so having a plugin is a great option after that. 

I'll also just add that the translation tests aren't failing because of flakiness, they are failing for a legitimate reason, that is being worked on in #6195.

Comment (dinhtungdu) [reply bin/wp-env-config.sh:24]: > I'll also just add that the translation tests aren't failing because of flakiness, they are failing for a legitimate reason, that is being worked on in https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/6195.

@alexflorisca Thanks for the head up! I update my test to install the language from the test. However, I don't think we need to uninstall the language, as long as we change the language back to English, i18n tests don't affect others. If we really want to uninstall the language files, we also need to remove the core, themes, and other plugins language files as well.

File: tests/e2e/specs/shopper/mini-cart.test.js [hunk 2]
```diff
@@ -559,4 +562,89 @@ describe( 'Shopper → Mini Cart', () => {
 			await expect( page ).toMatch( productTitle );
 		} );
 	} );
+
+	describe( 'Translated Mini Cart', () => {
```
Comment (alexflorisca) [inline tests/e2e/specs/shopper/mini-cart.test.js]: I've updated the E2E P2 pca54o-38z-p2 to indicate a standard way to write the `describe` and `it` blocks so it's consistent with the rest of the tests. Would be great to follow this here too.

File: tests/e2e/specs/shopper/mini-cart.test.js [hunk 3]
```diff
@@ -559,4 +562,89 @@ describe( 'Shopper → Mini Cart', () => {
...
+			await merchant.changeLanguage( 'nl_NL' );
+			await shopper.block.emptyCart();
+		} );
+
+		afterAll( async () => {
+			await merchant.changeLanguage( '' );
+		} );
+
+		describe( 'Classic Themes', () => {
+			useTheme( 'storefront' );
```
Comment (alexflorisca) [inline tests/e2e/specs/shopper/mini-cart.test.js]: I think this is unnecessary as. the default theme for all tests is storefront

**Review by alexflorisca — COMMENTED** [review]
Nice works, just a couple of comments. I think there might also have been a misunderstanding. I was more than happy with your approach of setting the language in `wp-config.sh` as it will speed up the tests like you said, especially when we now have multiple tests checking translations. It's ok to leave as it is for now and we can optimise later!

Main thing to address is just the standard way of writing the describe and it blocks so we are consistent with the rest of the tests, otherwise we're looking good 👍

File: tests/e2e/specs/shopper/mini-cart.test.js (hunk 2 above)
Comment (dinhtungdu) [reply tests/e2e/specs/shopper/mini-cart.test.js]: I don't have access to that P2, can you share a summary here?

Comment (nielslange) [reply tests/e2e/specs/shopper/mini-cart.test.js]: > I don't have access to that P2, can you share a summary here?

@dinhtungdu Besides the P2, Alex mentioned before, you can also find the proposed standard for writing `describe()` and `it()` messages on https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/6206. 

**TL;DR:**

```
describe(`Shopper > {Block name} > {Section (optional)}`) { ... }
```
```
it('user can {whatever is being tested}') { ... }
```

Comment (alexflorisca) [reply tests/e2e/specs/shopper/mini-cart.test.js]: Sure thing.

> `describe` and `it` blocks should use the following naming structure:

 ```js
describe('Shopper/Merchant → <BLOCK NAME> → (optional)<SECTION>', () =>{  
  (optional) describe('<SECTION', () => {
    it('User can <ACTION BEING TESTED>', () => {});
  })
})
```

for example: 

```js
describe('Shopper → Checkout', () =>{
  describe( 'Shipping and Billing Addresses', () => {
    it('User can have different shipping and billing addresses', () => {});
  });
})
```

or:

```js
describe('Shopper → Cart & Checkout → Translations', () =>{
  it('User can view translated Cart block', () => {});
})
```

**Comment by dinhtungdu** [comment]
> I was more than happy with your approach of setting the language in wp-config.sh as it will speed up the tests like you said, especially when we now have multiple tests checking translations. It's ok to leave as it is for now and we can optimise later!

@alexflorisca We don't need to rush merging this PR so I think we can spend time finding the best solution right from this PR.

I also thought about testing speed. If we have language files installed in a previous test, running WP CLI language commands in other tests has a slight impact on testing speed because the language has been downloaded already. In my tests on localhost, it took about 3s to run the command. On CI it may take longer.

Installing language in `wp-env-config.sh`, besides breaking isolation as you mentioned above, has another issue is, the language files are downloaded even when running a single test regardless of whether that test uses another language or not. However, it only affects local development, because, on CI, we run the entire test suite.

**Comment by alexflorisca** [comment]
> In my tests on localhost, it took about 3s to run the command. On CI it may take longer.

This will be significant if we add translation tests for other blocks, we are shaving 3s each time we add a test if we do it in setup file. Which I think is worth it. The E2E tests already take a long time to run so any optimisations we can find we should take advantage of

> the language files are downloaded even when running a single test regardless of whether that test uses another language or not. However, it only affects local development, because, on CI, we run the entire test suite.

I know @nielslange was working on separating the setup and teardown steps, so that when we run tests locally, we can run the setup file one, run our tests repeatedly until we are happy, then run the teardown. So this might not be such an issue after all.

**Comment by nielslange** [comment]
> I know @nielslange was working on separating the setup and teardown steps, so that when we run tests locally, we can run the setup file one, run our tests repeatedly until we are happy, then run the teardown. So this might not be such an issue after all.

Yes, I've indeed planned to decouple the setup and the teardown from the test runs. However, I cannot say when I'll be able to look into this. I'm afraid, that's nothing I'll do soon.

**Comment by dinhtungdu** [comment]
> This will be significant if we add translation tests for other blocks, we are shaving 3s each time we add a test if we do it in setup file. Which I think is worth it. The E2E tests already take a long time to run so any optimisations we can find we should take advantage of

@alexflorisca Makes a lot of sense to me. I reverted my changes to install language files from `wp-env-config.sh`.

File: bin/wp-env-config.sh [hunk 4]
```diff
@@ -20,4 +20,13 @@ wp plugin list
 wp theme activate storefront
 wp wc customer update 1 --user=1 --billing='{"first_name":"John","last_name":"Doe","company":"Automattic","country":"US","address_1":"addr 1","address_2":"addr 2","city":"San Francisco","state":"CA","postcode":"94107","phone":"123456789"}' --shipping='{"first_name":"John","last_name":"Doe","company":"Automattic","country":"US","address_1":"addr 1","address_2":"addr 2","city":"San Francisco","state":"CA","postcode":"94107","phone":"123456789"}'
 
+## Prepare translation for the test suite
+wp language core install nl_NL
+wp language plugin install woocommerce nl_NL
```
Comment (alexflorisca) [inline bin/wp-env-config.sh]: I don't think we need to install the language packs for `woocommerce`, `storefront` or `twentytwenty` here. The tests shouldn't be testing core woocommerce or theme functionality, they should only test translations that come from the woocommerce blocks repo.

File: tests/e2e/specs/shopper/mini-cart.test.js [hunk 5]
```diff
@@ -559,4 +562,77 @@ describe( 'Shopper → Mini Cart', () => {
...
 	} );
+
+	describe( 'Shopper → Mini Cart → Translations', () => {
+		beforeAll( async () => {
+			await merchant.changeLanguage( 'nl_NL' );
+			await shopper.block.emptyCart();
+		} );
+
+		afterAll( async () => {
+			await merchant.changeLanguage( '' );
```
Comment (alexflorisca) [inline tests/e2e/specs/shopper/mini-cart.test.js:572]: I'm not 100% sure this will revert back to english, best to be specific here an chage to `'en_EN'`.

**Review by alexflorisca — COMMENTED** [review]
Just a couple more comments @dinhtungdu and we can merge. Good work and hope the back and forth isn't too frustrating :)

File: tests/e2e/specs/shopper/mini-cart.test.js (hunk 5 above)
Comment (dinhtungdu) [reply tests/e2e/specs/shopper/mini-cart.test.js:572]: @alexflorisca The `English (United States)` option of the language dropdown has the value set to nothing. That's why I use `''` here. `en_EN` option actually doesn't exist.

File: bin/wp-env-config.sh (hunk 4 above)
Comment (dinhtungdu) [reply bin/wp-env-config.sh]: I thought, when changing language, WP will install languages for active plugins and theme, that's why I do it here to speed up the test. But you're right, WP only installs language for core when we change the site language to a new one.

**Comment by alexflorisca** [comment]
Nice, thanks for the changes @dinhtungdu! The `mini-cart.test.js` is failing in CI (but passes locally), so I wonder if we need to investigate that. I noticed that in this latest CI run, the mini cart ran as the first test, where as when I ran it locally, it ran after other tests. I wonder if we're making any assumptions or relying on some setup from other tests?

**Comment by dinhtungdu** [comment]
> The `mini-cart.test.js` is failing in CI (but passes locally), so I wonder if we need to investigate that.

Interestingly, the Mini Cart test fails on the `WP latest with Gutenberg plugin` only. It also didn't fail in the previous [commit](https://github.com/woocommerce/woocommerce-gutenberg-products-block/actions/runs/2148392957). Look like we're having flaky tests.

**Comment by dinhtungdu** [comment]
@alexflorisca Mini Cart translations tests are passing on both environments. Can you please take another look at this?

**Review by alexflorisca — APPROVED** [review]
Looks good, let's 🚢  it!
